ocaml/xapi/xapi_host_driver.ml

```diff
@@ -0,0 +1,297 @@
+               (set, map)
+           )
+           (DriverSet.empty, DriverMap.empty)
+    in
+    let filesystem_drivers =
+      Array.fold_left
+        (fun set name -> DriverSet.add name set)
+        DriverSet.empty
+        (Sys.readdir driver_paths.all_versions_dir)
+    in
+    (* Modify drivers that were in the DB before and are still on the
+       filesystem in-place to keep references from outside intact *)
+    let updated_drivers =
+      DriverSet.inter db_drivers_set filesystem_drivers
+      |> DriverSet.elements
+      |> List.filter_map (fun driver_name ->
+             let self = DriverMap.find driver_name db_drivers_map in
+             let selected_version, active_version, driver_versions =
+               get_driver_versions driver_name driver_paths.selected_dir
+                 driver_paths.all_versions_dir
+             in
+             let selected_changed =
+               Db.Host_driver.get_selected_version ~__context ~self
+               <> selected_version
+             in
+             let active_changed =
+               Db.Host_driver.get_active_version ~__context ~self
+               <> active_version
+             in
+             let versions_changed =
+               Db.Host_driver.get_versions ~__context ~self <> driver_versions
+             in
+             if selected_changed then
+               Db.Host_driver.set_selected_version ~__context ~self
+                 ~value:selected_version ;
+             if active_changed then
+               Db.Host_driver.set_active_version ~__context ~self
+                 ~value:active_version ;
+             if versions_changed then
+               Db.Host_driver.set_versions ~__context ~self
+                 ~value:driver_versions ;
+             if selected_changed || active_changed || versions_changed then
+               Some driver_name
+             else
+               None
+         )
+    in
+    (* Remove drivers that are no longer in the filesystem from the DB *)
+    let removed_drivers =
+      DriverSet.diff db_drivers_set filesystem_drivers
+      |> DriverSet.elements
+      |> List.map (fun driver_name ->
+             let self = DriverMap.find driver_name db_drivers_map in
+             destroy ~__context ~self ; driver_name
+         )
+    in
+    (* Add the new drivers *)
+    let added_drivers =
+      DriverSet.diff filesystem_drivers db_drivers_set
+      |> DriverSet.elements
+      |> List.map (fun driver_name ->
+             let selected_version, active_version, driver_versions =
+               get_driver_versions driver_name driver_paths.selected_dir
+                 driver_paths.all_versions_dir
+             in
+             create ~__context ~host ~name_label:driver_name
+               ~versions:driver_versions ~active_version ~selected_version ;
+             driver_name
+         )
+    in
+    [
+      ("updated_drivers", updated_drivers)
+    ; ("removed_drivers", removed_drivers)
+    ; ("added_drivers", added_drivers)
```

Comment:
I think it is better to log these changes rather than return them to the API caller, which in the common case would be internal (dbsync).

Reply:
Yeah. It seems to me that the data in xapi db is a read-only cache of the situation on the filesystem. E.g. even there is inconsistency between xapi db and filesystem, the xapi db will be updated according to the filesystem always. 
So would a much simpler approach be just to flush the cache completely and refill it, given there will not be too many such drivers?

Reply:
> So would a much simpler approach be just to flush the cache completely and refill it, given there will not be too many such drivers?

If we flush the cache and re-create it, we also invalidate all references to drivers that an API client could hold. I would like to avoid that.